packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,444 @@
+      }
+      count += currentString.length;
+    }
+    return 0;
+  }
+
+  /// Return the offset at the end of the nearest word to the right of the given
+  /// offset.
+  static int getRightByWord(String text, TextMetrics textMetrics, int offset, [bool includeWhitespace = true]) {
+    // If the selection is already all the way right, there is nothing to do.
+    if (offset == text.length) {
+      return offset;
+    }
+
+    // If we can just return the end of the text without checking for a word.
+    if (offset == text.length - 1 || offset == text.length) {
+      return text.length;
+    }
+
+    final int startPoint = includeWhitespace || !isWhitespace(text.codeUnitAt(offset))
+        ? offset
+        : nextCharacter(offset, text, includeWhitespace);
+    final TextRange nextWord = textMetrics.getWordBoundary(TextPosition(offset: startPoint));
+    return nextWord.end;
+  }
+
+  /// Creates a copy of this value but with the given fields replaced with the new values.
+  TextEditingValue copyWith({
+    String? text,
+    TextSelection? selection,
+    TextRange? composing,
+  }) {
+    return TextEditingValue(
+      text: text ?? this.text,
+      selection: selection ?? this.selection,
+      composing: composing ?? this.composing,
+    );
+  }
+
+  /// Whether the [composing] range is a valid range within [text].
+  ///
+  /// Returns true if and only if the [composing] range is normalized, its start
+  /// is greater than or equal to 0, and its end is less than or equal to
+  /// [text]'s length.
+  ///
+  /// If this property is false while the [composing] range's `isValid` is true,
+  /// it usually indicates the current [composing] range is invalid because of a
+  /// programming error.
+  bool get isComposingRangeValid => composing.isValid && composing.isNormalized && composing.end <= text.length;
+
+  // Deletes the current non-empty selection.
+  //
+  // If the selection is currently non-empty, this method deletes the selected
+  // text. Otherwise this method does nothing.
+  TextEditingValue _deleteNonEmptySelection() {
+    assert(selection.isValid);
+    assert(!selection.isCollapsed);
+
+    final String textBefore = selection.textBefore(text);
+    final String textAfter = selection.textAfter(text);
+    final TextSelection newSelection = TextSelection.collapsed(
+      offset: selection.start,
+      affinity: selection.affinity,
+    );
+    final TextRange newComposingRange = !composing.isValid || composing.isCollapsed
+      ? TextRange.empty
+      : TextRange(
+        start: composing.start - (composing.start - selection.start).clamp(0, selection.end - selection.start),
+        end: composing.end - (composing.end - selection.start).clamp(0, selection.end - selection.start),
+      );
+
+    return TextEditingValue(
+      text: textBefore + textAfter,
+      selection: newSelection,
+      composing: newComposingRange,
+    );
+  }
+
+  /// Deletes to the given index.
```

Comment:
I've changed it to say "inclusive" but tell me if I'm wrong here...

`wo|rd`

deleteTo(0) => `|rd`
deleteTo(4) => `wo|`

Is that right?  I think it confuses me because I'm thinking of the indices as the positions between the characters, not the characters themselves.

Reply:
Yeah that's my understanding as well.